gnome-initial-setup/demo/eos-demo-video.c

```diff
@@ -0,0 +1,230 @@
+/* -*- mode: C; c-file-style: "gnu"; indent-tabs-mode: nil; -*- */
+/*
+ * Copyright (C) 2017 Endless Mobile, Inc.
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License as
+ * published by the Free Software Foundation; either version 2 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, see <http://www.gnu.org/licenses/>.
+ *
+ * Written by:
+ *   Sam Spilsbury <sam@endlessm.com>
+ */
+
+#include "config.h"
+
+#include <eos-demo-video.h>
+
+typedef struct _GisDemoVideoPrivate
+{
+  GtkWindow *main_window;
+  GstElement *element;
+  GstElement *player;
+  GtkWidget *widget;
+  GstBus *bus;
+} GisDemoVideoPrivate;
+
+G_DEFINE_TYPE_WITH_PRIVATE (GisDemoVideo, gis_demo_video, GTK_TYPE_APPLICATION)
+
+static void
+on_demo_video_event (GtkWidget *widget,
+                     GdkEvent  *event,
+                     gpointer  user_data)
+{
+  if (event->type == GDK_BUTTON_PRESS ||
+      event->type == GDK_2BUTTON_PRESS ||
+      event->type == GDK_3BUTTON_PRESS ||
+      event->type == GDK_KEY_PRESS ||
+      event->type == GDK_MOTION_NOTIFY)
+    {
+      gtk_widget_destroy (widget);
+    }
+}
+
+static gboolean
+on_video_bus_event (GstBus     *bus,
+                    GstMessage *msg,
+                    gpointer   user_data)
+{
+  GisDemoVideo *demo_video = GIS_DEMO_VIDEO (user_data);
+  GisDemoVideoPrivate *priv = gis_demo_video_get_instance_private (demo_video);
+
+  switch (GST_MESSAGE_TYPE (msg))
+    {
+    case GST_MESSAGE_EOS:
+      gst_element_set_state (priv->player, GST_STATE_PAUSED);
+      gst_element_seek_simple (priv->player,
+                               GST_FORMAT_TIME,
+                               GST_SEEK_FLAG_FLUSH |
+                               GST_SEEK_FLAG_ACCURATE,
+                               0);
+      gst_element_set_state (priv->player, GST_STATE_PLAYING);
+    default:
+      break;
+    }
+
+  return G_SOURCE_CONTINUE;
+}
+
+#define DEMO_VIDEO_FILE_URI "file://" DEMOVIDEODATADIR "/eos-demo-video.ogv"
+
+/* The shell hides all windows on startup, by showing
+ * the overview which is understandable, so show the
+ * window a few seconds later */
```

Comment:
I thought this might catch your attention ;-)

It happens for any toplevel or fullscreen windows that are open when the shell starts up, because we show the icon grid on startup.

Now, personally, I believe that showing the icon grid on startup when there are already windows open is a bug (and somewhat annoying to people like myself who frequently `gnome-shell --replace`, though I gather that I am not the typical user ;-) ). I see two ways of dealing with this:

(1) If any windows are present when the shell starts up, don't automatically show the icon grid
(2) Continue to use the timeout to show the window a little bit later.

I don't think it makes sense to add a special case for fullscreen windows or for the demo splash video, since this problem is more general in nature. For instance, you could have another auto-starting application which happens to get launched and presented before the shell shows the icon grid and then it will get hidden by the shell shortly afterwards. So I'm leaning towards (1). Thoughts?